alibi_detect/cd/tensorflow/classifier.py

```diff
@@ -176,8 +176,13 @@ def score(self, x: np.ndarray) -> Tuple[float, float, np.ndarray, np.ndarray,  #
             else:
                 raise TypeError(f'x needs to be of type np.ndarray or list and not {type(x)}.')
             ds_tr = self.dataset(x_tr, y_tr)
-            self.model = clone_model(self.original_model) if self.retrain_from_scratch \
-                else self.model
+            if self.retrain_from_scratch:
+                # clone model to re-initialise
+                self.model = clone_model(self.original_model)
+                # If a new (tf>=2.11) optimizer, clone this too to prevent error due to cloned model
+                optimizer = self.train_kwargs['optimizer']
+                if isinstance(optimizer, tf.keras.optimizers.Optimizer):  # If not a legacy optimizer
```

Comment:
Thanks, you're right this didn't make sense. I haven't found a sensible backward compatible way to check new vs old optimizers. The `tf.keras.optimizers.legacy` and `tf.keras.optimizers.experimental` submodules don't exist in much older tf versions, and it isn't clear what will happen to these in the future anyway. To keep things simple I've just removed the check so that we clone old optimizers too. I've just checked with tf 2.2.0 and it works OK.